### PR TITLE
[FIX] Native FSDP precision + tests

### PR DESCRIPTION
**Pull request by SeanNaren**
## What does this PR do?

Related #12334 #12447.

After merging the above-associated PR and trying to use Native FSDP, I noticed there were many things wrong.

1. Tests were not being run as GPU CI uses 1.8.x PyTorch (LTS). This is an issue that is being resolved by potentially introducing an additional GPU run for PyTorch Latest. This meant the tests were not running, and they were all failing
2. 1.11 FSDP seems to be broken in various ways, with state_dict saving/loading issues + no mixed precision. So many fixes have come out for 1.12 (nightly) that no user really should be using 1.11, and hopefully, be using 1.12.
3. The native precision plugin was not selected when using Native FSDP.

This PR addresses 2 & 3, however, 1 remains a separate issue. I've run each test individually to ensure the integration works (as well as updated it a bit, and introduced mixed-precision support which was missing). I've also moved the requirement to 1.12dev, which will work for PyTorch Nightly.

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Review by awaelchli — APPROVED** [review]

**Comment by SeanNaren** [comment]
After some thought, I think I've reached a consensus:

Let's wait for 1.12 to be released, and revisit this PR when tests start failing (if we GPU test latest). Given past experience with BFloat16, I think its best to hold off trying to stabilize the API till the major release is out. If people end up trying to use FSDP native, we can point them to this PR for a fixed version.

**Review by Borda — APPROVED** [review]

**Comment by sisilmehta2000** [comment]
Most of the code is gated by the PYTORCH_1_12 flag. So then should we just land this before the PyTD 1.12 is released? cc @SeanNaren

File: tests/tests_pytorch/strategies/test_ddp_fully_sharded_native.py [hunk 1]
```diff
@@ -31,29 +32,27 @@ def test_invalid_on_cpu(tmpdir):
...
-            fast_dev_run=True,
-            strategy="fsdp_native",
-            accelerator="gpu",
-            devices=1,
-            precision=16,
-        )
-        assert isinstance(trainer.strategy, DDPFullyShardedNativeStrategy)
+@pytest.mark.parametrize("precision, expected", [(16, torch.float16), ("bf16", torch.bfloat16)])
+def test_precision_plugin_config(precision, expected):
+    plugin = FullyShardedNativeMixedPrecisionPlugin(precision=precision, device="cuda")
```
Comment (rohan-varma) [inline tests/tests_pytorch/strategies/test_ddp_fully_sharded_native.py:38]: Do we have a way to configure only parameters or reduce_dtype? 

i.e., a popular mixed precision approach is to set mixed precision for gradient communication (`reduce_dtype`) only. The config for this would look like:

```
MixedPrecision(reduce_dttype=bfloat16)
```

and param_dtype, config_dtype won't be touched or casted. Could we add support for this as well as this is basically gradient compression?

**Review by rohan-varma — COMMENTED** [review]
Wonder if there is a way for user to pass in all FSDP arguments that the ctor takes as of 1.12: https://github.com/pytorch/pytorch/blob/master/torch/distributed/fsdp/fully_sharded_data_parallel.py#L631

Currently, looks like we have CPU offload and now mixed precision added here, but curious about the other configs as well.

**Comment by SeanNaren** [comment]
After this, I suggest we do a docs overhaul for the FSDP logic. It's very out dated and can be hard to figure out exactly the right approach.

**Comment by carmocca** [comment]
@rohan-varma Absolutely! This is a high priority for the PL 1.7 release

File: src/pytorch_lightning/strategies/fully_sharded_native.py [hunk 2]
```diff
@@ -66,6 +71,8 @@ def __init__(  # type: ignore[no-untyped-def]
         process_group_backend: Optional[str] = None,
         cpu_offload=None,
         backward_prefetch=None,
+        mixed_precision=None,
```
Comment (rohan-varma) [inline src/pytorch_lightning/strategies/fully_sharded_native.py:74]: we may want to consider the story around checkpointing as well - there are currently 3 different APIs and some configuration that we support, it would be good to include this in lightning as well.

**Comment by SeanNaren** [comment]
As discussed with @carmocca offline, we should get this PR in even if we currently do not have 1.12 tests running on the GPU.

I have manually confirmed all tests pass using 1.12.

File: src/pytorch_lightning/strategies/fully_sharded_native.py [hunk 3]
```diff
@@ -109,16 +121,32 @@ def __init__(  # type: ignore[no-untyped-def]
...
         return self.parallel_devices[self.local_rank]
 
+    @property
+    def num_nodes(self) -> int:
+        return self._num_nodes
+
+    @num_nodes.setter
+    def num_nodes(self, num_nodes: int) -> None:
+        # note that world ranks is related to num_nodes, when resetting it, need to reset world ranks
+        self._num_nodes = num_nodes
```
Comment (carmocca) [inline src/pytorch_lightning/strategies/fully_sharded_native.py:144]: Why is this a getter/setter instead of the plain attribute if we are not customizing anything?

File: src/pytorch_lightning/trainer/connectors/accelerator_connector.py [hunk 4]
```diff
@@ -700,17 +700,13 @@ def _check_and_init_precision(self) -> PrecisionPlugin:
...
-                raise MisconfigurationException(
-                    "DDPFullyShardedNativeStrategy currently doesn't support Mixed Precision"
-                )
 
             if self._amp_type_flag == AMPType.NATIVE:
                 device = "cpu" if self._accelerator_flag == "cpu" else "cuda"
 
                 if isinstance(self.strategy, (DDPShardedStrategy, DDPSpawnShardedStrategy)):
                     return ShardedNativeMixedPrecisionPlugin(self._precision_flag, device)
-                if isinstance(self.strategy, DDPFullyShardedStrategy):
```
Comment (carmocca) [inline src/pytorch_lightning/trainer/connectors/accelerator_connector.py:713]: I'm confused here. Why did `DDPFullyShardedStrategy` choose the Native mixed precision plugin before?

File: src/pytorch_lightning/strategies/fully_sharded_native.py [hunk 5]
```diff
@@ -35,18 +37,23 @@
...
     from torch.distributed.fsdp.fully_sharded_data_parallel import (
         BackwardPrefetch,
         CPUOffload,
         FullyShardedDataParallel,
+        MixedPrecision,
     )
     from torch.distributed.fsdp.wrap import enable_wrap
-
+else:
+    MixedPrecision = None
```
Comment (rohitgr7) [inline src/pytorch_lightning/strategies/fully_sharded_native.py:54]:
```suggestion
    FullyShardedDataParallel = None
    MixedPrecision = None
```

File: src/pytorch_lightning/plugins/precision/fully_sharded_native_amp.py [hunk 6]
```diff
@@ -29,3 +38,18 @@ def clip_grad_by_norm(self, *_: Any, **__: Any) -> None:
...
+
+    @property
+    def mixed_precision_config(self) -> Optional[MixedPrecision]:
+        assert MixedPrecision is not None
+        if self.precision == PrecisionType.HALF:
+            dtype = torch.float16
+        elif self.precision == PrecisionType.BFLOAT:
+            dtype = torch.bfloat16
+        else:
+            raise MisconfigurationException(f"Was unable to infer precision type, received {self.precision}.")
```
Comment (rohitgr7) [inline src/pytorch_lightning/plugins/precision/fully_sharded_native_amp.py:50]:
```suggestion
            raise MisconfigurationException(f"Was unable to infer precision type, received {self.precision!r}.")
```

File: src/pytorch_lightning/strategies/fully_sharded_native.py [hunk 7]
```diff
@@ -84,22 +93,27 @@ def __init__(  # type: ignore[no-untyped-def]
...
+            backward_prefetch:
                 This is an experimental feature that is subject to change in the
                 the near future. It allows users to enable two different backward_prefetch
                 algorithms to help backward communication and computation overlapping.
                 Pros and cons of each algorithm is explained in the class ``BackwardPrefetch``.
+            mixed_precision:
+                Mixed Precision config. By default, Lightning will enable FP16 if ``precision=16`
+                or BF16 if ``precision=bf16`` unless a config is passed in.
+                This is only available in PyTorch 1.12 and later.
+            **kwargs: Passed to the FSDP Context manager which will configure the FSDP class when wrapping modules.
```
Comment (rohitgr7) [inline src/pytorch_lightning/strategies/fully_sharded_native.py:113]:
```suggestion
            \**kwargs: Passed to the FSDP Context manager which will configure the FSDP class when wrapping modules.
```

File: src/pytorch_lightning/strategies/fully_sharded_native.py [hunk 8]
```diff
@@ -84,22 +93,27 @@ def __init__(  # type: ignore[no-untyped-def]
...
         """
-        if not _TORCH_GREATER_EQUAL_1_11:
-            raise MisconfigurationException("DDPFullyShardedNativeStrategy is supported from pytorch v1.11.0 onwards.")
+        if not _TORCH_GREATER_EQUAL_1_12:
+            raise MisconfigurationException("DDPFullyShardedNativeStrategy is supported from pytorch v1.12.0 onwards.")
```
Comment (rohitgr7) [inline src/pytorch_lightning/strategies/fully_sharded_native.py:116]:
```suggestion
            raise MisconfigurationException("`DDPFullyShardedNativeStrategy` is supported from PyTorch v1.12.0 onwards.")
```

File: src/pytorch_lightning/strategies/fully_sharded_native.py (hunk 5 above)
Comment (rohitgr7) [reply src/pytorch_lightning/strategies/fully_sharded_native.py:54]: might need to add assertions somewhere for mypy

Comment (carmocca) [reply src/pytorch_lightning/strategies/fully_sharded_native.py:54]: mypy is currently happy. Why do we need this?
